zebra-consensus/src/verify/block.rs

```diff
@@ -171,4 +289,186 @@ mod tests {
         }
     }
+
+    #[tokio::test]
+    #[spandoc::spandoc]
+    async fn verify() -> Result<(), Report> {
+        install_tracing();
+
+        let block =
+            Arc::<Block>::zcash_deserialize(&zebra_test::vectors::BLOCK_MAINNET_415000_BYTES[..])?;
+        let hash: BlockHeaderHash = block.as_ref().into();
+
+        let state_service = Box::new(zebra_state::in_memory::init());
+        let mut block_verifier = super::init(state_service);
+
+        /// Make sure the verifier service is ready
+        let ready_verifier_service = block_verifier.ready_and().await.map_err(|e| eyre!(e))?;
+        /// Verify the block
+        let verify_response = ready_verifier_service
+            .call(block.clone())
+            .await
+            .map_err(|e| eyre!(e))?;
+
+        assert_eq!(verify_response, hash);
+
+        Ok(())
+    }
+
+    #[tokio::test]
+    #[spandoc::spandoc]
+    async fn round_trip() -> Result<(), Report> {
+        install_tracing();
+
+        let block =
+            Arc::<Block>::zcash_deserialize(&zebra_test::vectors::BLOCK_MAINNET_415000_BYTES[..])?;
+        let hash: BlockHeaderHash = block.as_ref().into();
+
+        let mut state_service = zebra_state::in_memory::init();
+        let mut block_verifier = super::init(state_service.clone());
+
+        /// Make sure the verifier service is ready
+        let ready_verifier_service = block_verifier.ready_and().await.map_err(|e| eyre!(e))?;
+        /// Verify the block
+        let verify_response = ready_verifier_service
+            .call(block.clone())
+            .await
+            .map_err(|e| eyre!(e))?;
+
+        assert_eq!(verify_response, hash);
+
+        /// Make sure the state service is ready
+        let ready_state_service = state_service.ready_and().await.map_err(|e| eyre!(e))?;
+        /// Make sure the block was added to the state
+        let state_response = ready_state_service
+            .call(zebra_state::Request::GetBlock { hash })
+            .await
+            .map_err(|e| eyre!(e))?;
+
+        if let zebra_state::Response::Block {
+            block: returned_block,
+        } = state_response
+        {
+            assert_eq!(block, returned_block);
+        } else {
+            bail!("unexpected response kind: {:?}", state_response);
+        }
+
+        Ok(())
+    }
+
+    #[tokio::test]
+    #[spandoc::spandoc]
+    async fn verify_fail_add_block() -> Result<(), Report> {
+        install_tracing();
+
+        let block =
+            Arc::<Block>::zcash_deserialize(&zebra_test::vectors::BLOCK_MAINNET_415000_BYTES[..])?;
+        let hash: BlockHeaderHash = block.as_ref().into();
+
+        let mut state_service = zebra_state::in_memory::init();
+        let mut block_verifier = super::init(state_service.clone());
+
+        /// Make sure the verifier service is ready (1/2)
+        let ready_verifier_service = block_verifier.ready_and().await.map_err(|e| eyre!(e))?;
```

Comment:
Oh, I see. The double await issue presents an issue here :( I'll start brain storming ways to support this better...

Reply:
A similar issue also turns up if you try to add a span to a parent block containing multiple awaits. Even if each await already has its own span. (For example, a loop containing multiple awaits.)

Reply:
aah yea, i figured that would happen based on the way I implemented it the actual instrumentation ontop of syn's expression visitor, it just iterates over statements within every block and for each statement that has a doc comment it tries to instrument the await with an await visitor, but it cant tell the difference between an await thats in a statement inside of a block inside of the current statement or if its an await thats directly part of the expression you instrumented. Not really sure how to fix it... maybe @mystor would have some ideas